src/zerocash/Coin.cpp

```diff
@@ -36,29 +22,17 @@ Coin::Coin(): addr_pk(), cm(), rho(ZC_RHO_SIZE), r(ZC_R_SIZE), coinValue(ZC_V_SI
 
 }
 
-Coin::Coin(const std::string bucket, Address& addr): addr_pk(), cm(), rho(ZC_RHO_SIZE), r(ZC_R_SIZE), k(ZC_K_SIZE), coinValue(ZC_V_SIZE) {
-    // Retreive and decode the private key
-    ECIES<ECP>::PrivateKey decodedPrivateKey;
-    decodedPrivateKey.Load(StringStore(addr.getPrivateAddress().getEncryptionSecretKey()).Ref());
-
-    // Create the decryption session
-    AutoSeededRandomPool prng;
-    ECIES<ECP>::Decryptor decrypt(decodedPrivateKey);
-
-    // Convert the input string into a vector of bytes
-    std::vector<byte> bucket_bytes(bucket.begin(), bucket.end());
-
-    // Construct a temporary object to store the plaintext, large enough
-    // to store the plaintext if it were extended beyond the real size.
-    std::vector<unsigned char> plaintext;
-    // Size as needed, filling with zeros.
-    plaintext.resize(decrypt.MaxPlaintextLength(decrypt.CiphertextLength(ZC_V_SIZE + ZC_R_SIZE + ZC_RHO_SIZE)), 0);
+Coin::Coin(const ZCNoteEncryption::Ciphertext& bucket,
+           Address& addr,
+           uint256& epk,
+           unsigned char nonce
+           ): addr_pk(), cm(), rho(ZC_RHO_SIZE), r(ZC_R_SIZE), k(ZC_K_SIZE), coinValue(ZC_V_SIZE) {
 
-    // Perform the decryption
-    decrypt.Decrypt(prng,
-                    &bucket_bytes[0],
-                    decrypt.CiphertextLength(ZC_V_SIZE + ZC_R_SIZE + ZC_RHO_SIZE),
-                    &plaintext[0]);
+    ZCNoteDecryption decrypter(addr.getPrivateAddress().getEncryptionSecretKey());
+    auto plaintext = decrypter.decrypt(bucket,
+                                       epk,
+                                       uint256(),
+                                       nonce);
```

Comment:
This can throw an `std::runtime_error` for a modified note ciphertext. Where's that case handled? Will it currently crash the daemon? If so, file a ticket to implement handling of that case.